### PR TITLE
Indexes S3 object metadata

### PR DESCRIPTION
**Pull request by t-davies**
**What**  
Indexes metadata on the S3 objects alongside the metadata from the metadata file.

**Why**  
For document uploads we need to be able to set a description of the document at the time of uploading it, this allows a file description to be added (or any other metadata for other, future, use cases) and indexed.

### CODE REVIEW

File: src/gateways/S3Gateway.ts [hunk 1]
```diff
@@ -75,4 +75,15 @@ export class S3Gateway {
...
+
+  async getObjectMetadata(key: string): Promise<Record<string, string>> {
+    const metadata = await this.client
+      .headObject({
+        Bucket: this.bucketName,
+        Key: key
+      })
+      .promise();
+
+    return metadata.Metadata ?? {};
```
Comment (kosiakkatrina) [inline src/gateways/S3Gateway.ts:87]: TIL, very nice

File: src/gateways/S3Gateway.test.ts [hunk 2]
```diff
@@ -103,4 +103,26 @@ describe('S3Gateway', () => {
...
+
+    const gateway = new S3Gateway({
+      logger: new NoOpLogger(),
+      client,
+      bucketName: 'testBucket'
+    });
+
+    const result = await gateway.getObjectMetadata('123/Passport.jpg');
+    expect(result).toStrictEqual(expectedObject);
+  });
```
Comment (kosiakkatrina) [inline src/gateways/S3Gateway.test.ts:127]: Maybe a test for when there's no metadata?

File: src/use-cases/GetMetadata.ts [hunk 3]
```diff
@@ -17,7 +18,18 @@ export default class GetMetadataUseCase implements UseCase<GetMetadataQuery, Doc
...
+      objectMetadata,
+      predefinedMetadata
+    ] = await Promise.all([
+      this.metadata.getObjectMetadata(objectKey),
+      this.metadata.get(documentId)
+    ]);
+
+    return {
+      ...objectMetadata,
+      ...predefinedMetadata
```
Comment (kosiakkatrina) [inline src/use-cases/GetMetadata.ts:32]: That's a nice syntax

File: src/gateways/S3Gateway.ts [hunk 4]
```diff
@@ -49,7 +49,8 @@ export class S3Gateway {
           Conditions: [
             { bucket: this.bucketName },
             ['starts-with', '$key', `${documentId}/`],
-            { 'X-Amz-Server-Side-Encryption': 'AES256' }
+            { 'X-Amz-Server-Side-Encryption': 'AES256' },
+            ['starts-with', '$X-Amz-Meta-Description', '']
```
Comment (kosiakkatrina) [inline src/gateways/S3Gateway.ts:53]: Wouldn't everything start with '' ? 🤔 Is it just making sure it's not null?

Comment (t-davies) [reply src/gateways/S3Gateway.ts:53]: Yep, this is just making sure the value is set, even empty strings are allowed - for cases where there's no description.

**Comment by t-davies** [comment]
Note: need to swap the get metadata endpoint to read the metadata from Elasticsearch and not from S3. Will wait until the other ES bits have been merged - this will fix the integration tests.

**Comment by t-davies** [comment]
Yeah good shout, I'd forgotten about the changes to instance size. I'll add those in! Thanks.

File: serverless.yml [hunk 5]
```diff
@@ -148,3 +138,16 @@ custom:
     basePath: ""
     certificateName: "*.api.hackney.gov.uk"
     createRoute53Record: false
+  elasticsearch:
```
Comment (mmmmillar) [inline serverless.yml:141]: nice

**Pull request merged**